conformance/tests/gateway-infrastructure.go

```diff
@@ -0,0 +1,113 @@
+		kubernetes.GatewayMustHaveLatestConditions(t, s.Client, s.TimeoutConfig, gwNN)
+
+		t.Logf("retrieving Gateway %s/%s", gwNN.Namespace, gwNN.Name)
+		currentGW := &v1.Gateway{}
+		err := s.Client.Get(ctx, gwNN, currentGW)
+		require.NoError(t, err, "error getting Gateway: %v", err)
+		t.Logf("verifying that the Gateway %s/%s is accepted with infrastructure declared", gwNN.Namespace, gwNN.Name)
+		kubernetes.GatewayMustHaveCondition(t, s.Client, s.TimeoutConfig, gwNN, metav1.Condition{
+			Type:   string(v1.GatewayConditionAccepted),
+			Status: metav1.ConditionTrue,
+		})
+
+		t.Logf("verifying that generated resources for Gateway %s/%s have the proper gateway name label", gwNN.Namespace, gwNN.Name)
+		// Don't check services because implementations may have special filtering logic (e.g. for LB annotations)
+		// Instead, check service accounts for the gateway-name label first and then
+		// fallback to Pod if that fails
+		annotations := make(map[string]string, len(currentGW.Spec.Infrastructure.Annotations))
+		labels := make(map[string]string, len(currentGW.Spec.Infrastructure.Labels))
+		// Need to translate from (Annotation|Label)Key to string
+		for k, v := range currentGW.Spec.Infrastructure.Annotations {
+			annotations[string(k)] = string(v)
+		}
+
+		for k, v := range currentGW.Spec.Infrastructure.Labels {
+			labels[string(k)] = string(v)
+		}
+		saList := corev1.ServiceAccountList{}
+		podList := corev1.PodList{}
+		serviceList := corev1.ServiceList{}
+		err = s.Client.List(ctx, &saList, client.MatchingLabels{"gateway.networking.k8s.io/gateway-name": gwNN.Name}, client.InNamespace(ns))
+		require.NoError(t, err, "error listing ServiceAccounts")
+		err = s.Client.List(ctx, &podList, client.MatchingLabels{"gateway.networking.k8s.io/gateway-name": gwNN.Name}, client.InNamespace(ns))
+		require.NoError(t, err, "error listing Pods")
+		err = s.Client.List(ctx, &serviceList, client.MatchingLabels{"gateway.networking.k8s.io/gateway-name": gwNN.Name}, client.InNamespace(ns))
+		require.NoError(t, err, "error listing Services")
+		if len(saList.Items) > 0 {
+			sa := saList.Items[0]
+			require.Subsetf(t, sa.Labels, labels, "expected Pod label set %v to contain all Gateway infrastructure labels %v", sa.Labels, labels)
+			require.Subsetf(t, sa.Annotations, annotations, "expected Pod annotation set %v to contain all Gateway infrastructure annotations %v", sa.Annotations, annotations)
+		} else if len(podList.Items) > 0 {
+			// Fallback to pod
+			pod := podList.Items[0]
+			require.Subsetf(t, pod.Labels, labels, "expected Pod label set %v to contain all Gateway infrastructure labels %v", pod.Labels, labels)
+			require.Subsetf(t, pod.Annotations, annotations, "expected Pod annotation set %v to contain all Gateway infrastructure annotations %v", pod.Annotations, annotations)
+		} else {
+			require.NotEmpty(t, serviceList.Items, 0, "expected at least one Pod, ServiceAccount, or Service with gateway-name label")
```

Comment:
The comment for this function says `NotEmpty asserts that the specified object is NOT empty. I.e. not nil, "", false, 0 or either a slice or a channel with len == 0` so since `serviceList` is guaranteed to be initialized, it's safe if `serviceList.Items` is nil

Reply:
ah the panic is coming from extra argument (i.e. 0)

```go
func TestDummy(t *testing.T) {
	serviceList := corev1.ServiceList{}
	require.NotEmpty(t, serviceList.Items, 0, "expected at least one Pod, ServiceAccount, or Service with gateway-name label")
}
...
=== RUN   TestDummy
--- FAIL: TestDummy (0.00s)
panic: interface conversion: interface {} is int, not string [recovered]
	panic: interface conversion: interface {} is int, not string

goroutine 32 [running]:
testing.tRunner.func1.2({0x2c30d00, 0xc00026fa40})
	/usr/lib/go/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/usr/lib/go/src/testing/testing.go:1634 +0x377
panic({0x2c30d00?, 0xc00026fa40?})
	/usr/lib/go/src/runtime/panic.go:770 +0x132
github.com/stretchr/testify/assert.messageFromMsgAndArgs({0xc0000f8f40?, 0x3157480?, 0x5464680?})
	/home/tammach/go/src/github.com/cilium/cilium/vendor/github.com/stretchr/testify/assert/assertions.go:292 +0xed
```